### PR TITLE
Fix/bitswap hang

### PR DESCRIPTION
**Pull request by whyrusleeping**
Bitswap has had a race condition type bug in it for a very long time now, and it manifests rarely. We've always just assumed it was a fluke, but i discovered a repro, and a fix!!

When dialing we dial a bunch of addresses at once, if we get two connections to a peer, one of them gets closed, but not before sending a disconnect notification. Bitswap receives this notification and assumes we are no longer connected to that peer, and closes the worker assigned to that peer, even though we have an active connection to them.

### CODE REVIEW

File: exchange/bitswap/wantmanager.go [hunk 1]
```diff
@@ -101,13 +103,13 @@ func (pm *WantManager) SendBlock(ctx context.Context, env *engine.Envelope) {
 }
 
 func (pm *WantManager) startPeerHandler(p peer.ID) *msgQueue {
-	_, ok := pm.peers[p]
+	mq, ok := pm.peers[p]
 	if ok {
-		// TODO: log an error?
+		mq.refcnt++
```
Comment (jbenet) [inline exchange/bitswap/wantmanager.go]: i strongly doubt this is threadsafe. did you run tests with `-race`?

Comment (jbenet) [inline exchange/bitswap/wantmanager.go]: unless the whole WantManager is run by one thread

Comment (jbenet) [inline exchange/bitswap/wantmanager.go]: ok nvm. resolved. ignore this.

**Comment by jbenet** [comment]
@whyrusleeping can you update iptb in a different commit?

**Comment by jbenet** [comment]
Nice discovery, this will be good to have fixed. LGTM, i think.

**Comment by jbenet** [comment]
(though all travis tests fail!)

**Comment by jbenet** [comment]
test/dependencies/iptb/main.go:23:2: cannot find package "github.com/jbenet/go-multiaddr-net/Godeps/_workspace/src/github.com/jbenet/go-multiaddr"

**Comment by whyrusleeping** [comment]
yeah, i messed up the godeps stuff... i'm fixing that now

**Pull request merged**